bigquery/bigquery_migration_quickstart/main.go

```diff
@@ -0,0 +1,137 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START bigquerymigration_quickstart]
+
+// The bigquery_migration_quickstart application demonstrates basic usage of the
+// BigQuery migration API by executing workflow that performs an offline SQL
+// translation task.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"time"
+
+	migration "cloud.google.com/go/bigquery/migration/apiv2alpha"
+	translationtaskpb "google.golang.org/genproto/googleapis/cloud/bigquery/migration/tasks/translation/v2alpha"
+	migrationpb "google.golang.org/genproto/googleapis/cloud/bigquery/migration/v2alpha"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+func main() {
+	// Define two command line flags for controlling the behavior of this quickstart.
+	projectID := flag.String("project_id", "", "Cloud Project ID.")
+	location := flag.String("location", "us", "BigQuery Migration location used for interactions.")
+
+	// Parse flags and do some minimal validation.
+	flag.Parse()
+	if *projectID == "" {
+		log.Fatal("empty --project_id specified, please provide a valid project ID")
+	}
+	if *location == "" {
+		log.Fatal("empty --location specified, please provide a valid location")
+	}
+
+	ctx := context.Background()
+	migClient, err := migration.NewClient(ctx)
+	if err != nil {
+		log.Fatalf("migration.NewClient: %v", err)
+	}
+	defer migClient.Close()
+
+	workflow, err := executeTranslationWorkflow(ctx, migClient, *projectID, *location)
+	if err != nil {
+		log.Fatalf("workflow execution failed: %v", err)
+	}
+
+	reportWorkflowStatus(workflow)
+}
+
+// executeTranslationWorkflow constructs a migration workflow that performs some offline SQL translation.
+func executeTranslationWorkflow(ctx context.Context, client *migration.Client, projectID, location string) (*migrationpb.MigrationWorkflow, error) {
+
+	// Tasks are extensible; the translation task is defined by the BigQuery Migration API.
+
+	translationDetails := &translationtaskpb.TranslationTaskDetails{
+		// TODO: we need to stage some suitable translation inputs and complete this definition.
+
+		InputPath: "",
+	}
+
+	// We then convert the task details for translation into the suitable protobuf `Any` representation needed
+	// to define the workflow.
+	anyDetails, err := anypb.New(translationDetails)
+	if err != nil {
+		return nil, err
+	}
+
+	// Finally, construct the workflow creation request.
+	req := &migrationpb.CreateMigrationWorkflowRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/%s", projectID, location),
+		MigrationWorkflow: &migrationpb.MigrationWorkflow{
+			DisplayName: "example SQL conversion",
+			Tasks: map[string]*migrationpb.MigrationTask{
+				"example_conversion": {
+					Type:    "Translation_Teradata",
+					Details: anyDetails,
```

Comment:
I'm a bit confused why the API doesn't use a oneof here. Presumably it only supports a limited number of types + detail messages.

Reply:
It's because the intent is to allow user extension; so you end up with an arbitrary set of task definitions without a central registry.